Improve PathFinder/TargetedMovementGenerator

Stoabrogga opened the pull request:
## 🍰 Pullrequest
1. Ensure that the Z coordinates of the calculated path are correct. This is similar to what Cmangos and Trinitycore do:
https://github.com/cmangos/mangos-classic/blob/e593dbcde632ffc4ba04e82ffc01a51695d578a8/src/game/MotionGenerators/PathFinder.cpp#L983
https://github.com/TrinityCore/TrinityCore/blob/7f5790db6701a46b06b2c3319c3776c1fb93e3d5/src/server/game/Movement/PathGenerator.cpp#L596
2. Improve TargetedMovementGenerator for following pets
3. Fix TerrainInfo::GetWaterOrGroundLevel (can create values below ground level)
4. Remove unused variable in MoveSplineInit.cpp

### Proof
see #2726

### Issues
fixes #2726

### How2Test
Run around with warlock or hunter pets.

### Todo / Checklist
none

Stoabrogga committed:
Improve PathFinder

ratkosrb commented:
Does this work correctly for flying and swimming mobs?

ShiyoKozuki commented:
> Does this work correctly for flying and swimming mobs?



Not sure if this is related but...
On classic, mobs "run away in fear" and will path into water and swim away (if they started on land before running away).
On vmangos, they will never path into water/swim when "run away in fearing"ing (if they started on land before running away).

ghost commented:
> Does this work correctly for flying and swimming mobs?

Flying or swimming creatures do not use pathfinding at all, they use shortcuts (2 point paths) and unlike Cmangos and TC I intentionally omitted the z correction there because it should only apply to calculated paths.

@ShiyoKozuki This is not related, you should open an issue if the creature behavior is wrong.

Stoabrogga committed:
Fix TerrainInfo::GetWaterOrGroundLevel

Fix returning z values which are below ground level (can happen in shallow water).

Stoabrogga committed:
Remove unused variable

Stoabrogga committed:
Improve TargetedMovementGenerator for following pets

Stoabrogga committed:
Update TerrainInfo::GetWaterOrGroundLevel

ghost commented:
Added a few other fixes concerning #2726, see first comment.

Stoabrogga committed:
Update TargetedMovementGenerator

Ensure that speed is not 0
